src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp

```diff
@@ -1325,11 +1326,40 @@ void CascadiaSettings::_resolveNewTabMenuProfilesSet(const IVector<Model::NewTab
         // For a folder, we simply call this method recursively
         case NewTabMenuEntryType::Folder:
         {
-            const auto folderEntry = entry.as<Model::FolderEntry>();
-            auto folderEntries = folderEntry.Entries();
+            // We need to access the unfiltered entry list, a field that is not exposed
+            // in the projected class. So, we need to first obtain our implementation struct
+            // instance, to access this field.
+            const auto folderEntry{ winrt::get_self<implementation::FolderEntry>(entry.as<Model::FolderEntry>()) };
+
+            auto folderEntries = folderEntry->RawEntries();
             _resolveNewTabMenuProfilesSet(folderEntries, remainingProfiles, remainingProfilesEntry);
             break;
         }
+
+        // For a "matchProfiles" entry, we iterate through the list of all profiles and
+        // find all those matching: generated by the same source, having the same name, or
+        // having the same commandline. This can be expanded with regex support in the future.
+        // We make sure that none of the matches are included in the "remaining profiles" section.
+        case NewTabMenuEntryType::MatchProfiles:
+        {
+            const auto matchEntry = entry.as<Model::MatchProfilesEntry>();
+            matchEntry.Profiles(single_threaded_map<int, Model::Profile>());
+
+            auto activeProfileCount = gsl::narrow_cast<int>(_activeProfiles.Size());
+            for (auto profileIndex = 0; profileIndex < activeProfileCount; profileIndex++)
+            {
+                const auto profile = _activeProfiles.GetAt(profileIndex);
+
+                // On a match, we store it in the entry and remove it from the remaining list
+                if ((!matchEntry.Source().empty() && profile.Source() == matchEntry.Source()) || (!matchEntry.Name().empty() && profile.Name() == matchEntry.Name()) || (!matchEntry.CommandLine().empty() && profile.Commandline() == matchEntry.CommandLine()))

```

Comment:
recommend moving this to its own method, possibly on the `impl` type for `MatchProfilesEntry` to avoid having to WinRT-ize it

for example: we won't want to put all the regex matching code in this function, long as it is, so it'll be best to get ahead of it with a "does this profile match this entry?" method

Reply:
Done, a lot better this way indeed.